File: test/bokeh_apps/slider_tool/slide.py
"""Slider tool"""
import numpy as np
import bokeh.models

JS_CODE = """
    /**
     *  Full JS slider implementation
     *     - Hide/show left images
     *     - Hide/show right images
     *     - Move vertical line
     */
    console.log('mouse move');
"""


class Slider(object):
    """Controls alpha values of bokeh RGBA ColumnDataSources

    Understands bokeh architecture, either a GlyphRenderer
    or a ColumnDataSource may be passed in for either
    left_images or right_images arguments

    The various properties needed to set alpha values relative
    to the mouse position can be found by inspecting
    either data structure

    .. note:: If a ColumnDataSource is passed then the following
              keys are used to find the image properties
              'image', 'x', 'y', 'dw', 'dh'

    :param left_images: ColumnDataSource or GlyphRenderer used to
                        define RGBA images to the left of the cursor
    :param right_images: ColumnDataSource or GlyphRenderer used to
                         define RGBA images to the right of the cursor
    """
    def __init__(self, left_images, right_images):
        self.left_images = left_images
        self.left_alpha = bokeh.models.ColumnDataSource({
            "alpha": self.get_alpha(self.left_images)
        })
        self.right_images = right_images
        self.right_alpha = bokeh.models.ColumnDataSource({
            "alpha": self.get_alpha(self.right_images)
        })
        self.span = bokeh.models.Span(location=0,
                                      dimension='height',
                                      line_color='black',
                                      line_width=1)
        shared = bokeh.models.ColumnDataSource({
            "first_time": [False],
            "previous_mouse_x": [0]
        })
        self.mousemove = bokeh.models.CustomJS(args=dict(
            span=self.span,
            left_images=self.left_images,
            left_alpha=self.left_alpha,
            right_images=self.right_images,
            right_alpha=self.right_alpha,
            shared=shared
        ), code=JS_CODE)

    @staticmethod
    def get_alpha(bokeh_obj):
        """Helper to copy alpha values from RGBA bokeh GlyphRenderer or ColumnDataSource"""
        if hasattr(bokeh_obj, "data_source"):
            renderer = bokeh_obj
            if isinstance(renderer.glyph.image, basestring):
                images = renderer.data_source[renderer.glyph.image]
            else:
                images = renderer.glyph.image
        else:
            column_data_source = bokeh_obj
            images = column_data_source.data["image"]
        return [np.copy(rgba[:, :, -1]) for rgba in images]

    def add_figure(self, figure):
        """Attach various callbacks to a particular figure"""
        figure.js_on_event("mousemove", self.mousemove)

File: test/bokeh_apps/slider_tool/slide.js
// bokeh.models.CustomJS callback code
//
// args:
//      span - bokeh.models.Span
//      left_images - ColumnDataSource
//      left_alpha - Copy of left_images alpha values
//      right_images - ColumnDataSource
//      right_alpha - Copy of right_images alpha values
//      shared - ColumnDataSource
// cb_data:
//      geometry.x - mouse x position relative to figure
//
let slide_image = function(source,
                           original_alpha,
                           mouse_x,
                           previous_mouse_x,
                           first_time,
                           shape,
                           side) {
    // Hard-coded values for now
    let x = 0;
    let dw = 10;

    // Shared data
    let ni = shape[0];
    let nj = shape[1];

    // Mouse position(s)
    let left_x;
    let right_x;
    if (!isFinite(mouse_x)) {
        return;
    }
    if (mouse_x > previous_mouse_x) {
        left_x = previous_mouse_x;
        right_x = mouse_x;
    } else {
        left_x = mouse_x;
        right_x = previous_mouse_x;
    }

    // Update alpha pseudo-2D and RGBA pseudo-3D arrays
    let pixel_x;
    let alpha;
    let alpha_index;
    let image_alpha_index;
    let dy = dw / nj;
    let skip = 0;
    for (let j=0; j<nj; j++) {
        pixel_x = x + (j * dy);

        // Optimised selection of columns between mouse events
        // note: feature turned off during first paint
        if (!first_time) {
            if ((pixel_x > right_x) || (pixel_x < left_x)) {
                // pixel outside current and previous mouse positions
                skip += 1;
                continue;
            }
        }

        // Ordinary loop logic
        for (let i=0; i<ni; i++) {
            alpha_index = (nj * i) + j;
            original_alpha_value = original_alpha[alpha_index];
            if (original_alpha_value == 0) {
                continue;
            }
            image_alpha_index = (4 * alpha_index) + 3;
            if (visible_pixel(pixel_x, mouse_x, side)) {
               alpha = original_alpha_value;
            } else {
               alpha = 0;
            }
           source.data["image"][0][image_alpha_index] = alpha;
        }
    }
    if (skip !== nj) {
        // Some columns need to be painted
        source.change.emit();
    }
};

let visible_pixel = function(pixel_x, mouse_x, show_side) {
    if (show_side === "left") {
        return pixel_x < mouse_x;
    } else {
        return pixel_x > mouse_x;
    }
};

// CustomJS callback main program
let main = function() {
    // Gather data from cb_data and args
    let first_time = shared.data.first_time[0];
    let previous_mouse_x = shared.data.previous_mouse_x[0];
    let mouse_x = cb_data.geometry.x;

    // Move vertical line to mouse position
    span.location = mouse_x;

    // Left image extra data
    let left_alpha = left_extra.data["alpha"][0];
    let left_shape = left_extra.data["shape"][0];

    // Update image alpha values
    slide_image(left_images,
                left_alpha,
                mouse_x,
                previous_mouse_x,
                first_time,
                left_shape,
                "left");

    // Right image extra data
    let right_alpha = right_extra.data["alpha"][0];
    let right_shape = right_extra.data["shape"][0];

    // Update image alpha values
    slide_image(right_images,
                right_alpha,
                mouse_x,
                previous_mouse_x,
                first_time,
                right_shape,
                "right");

    // Update shared data
    shared.data.previous_mouse_x[0] = mouse_x;
    shared.data.first_time[0] = false;
};

main();
